### ISSUE TITLE
HMF search for degree=2 and dimension=2 times out on prod (but not on beta)

### ISSUE THREAD

**Issue by AndrewVSutherland** [LMFDB/lmfdb#1599]
The search http://www.lmfdb.org/ModularForm/GL2/TotallyReal/?field_label=&deg=2&disc=&weight=&level_norm=&dimension=2&count=100 times out after 30 seconds with a server error.  But the same search on beta http://beta.lmfdb.org/ModularForm/GL2/TotallyReal/?field_label=&deg=2&disc=&weight=&level_norm=&dimension=2&count=100 works fine, and on prod searching for just dimension=2 or just degree=2 works fine.

http://www.lmfdb.org/ModularForm/GL2/TotallyReal/?field_label=&deg=&disc=&weight=&level_norm=&dimension=2&count=100
http://www.lmfdb.org/ModularForm/GL2/TotallyReal/?field_label=&deg=2&disc=&weight=&level_norm=&dimension=&count=100

Reported by Marc Masdeu at Warwick (respond via e-mail when fixed).

**Comment by AndrewVSutherland** [comment]
I am adding an index to better support this kind of query; HMF searches may be unavailable for the next few minutes while the index builds.

**Comment by AndrewVSutherland** [comment]
@edgarcosta I have done some testing and I believe this is may be a performance issue caused by the switch to wired tiger zlib.  I just recopied all the HMF data over from Warwick so the hmfs.forms collection in the two dbs are identical, and they have the exact same indexes, in the same order.

The issue is that when performing a query that cannot be resolved entirely by index lookup, some subset of the records in the collection will need to be scanned, and in the case of the forms collection, the average size of these records is quite large (about 200kb) and my guess is that with zlib each of these is getting decompressed (even though only one or two attributes needs to be examined) and this takes a lot of time.

Here is a timing comparison between a query run on atkin and a query run on www.cental0.lmfdb.xyz that I believe reflects what mongo db is doing internally to resolve this query (it searches the list of records with dimension 2 to figure out which also have degree 2):

On Atkin:

time A=[(r['label'],r['deg']) for r in db.forms.find({'dimension':int(2)},{'label':True,'deg':True})]
CPU times: user 607 ms, sys: 48 ms, total: 655 ms

On www.central0.lmfdb.xyz:

time A=[(r['label'],r['deg']) for r in db.forms.find({'dimension':int(2)},{'label':True,'deg':True})]
CPU times: user 464 ms, sys: 44 ms, total: 508 ms
Wall time: 4min 54s

Don't be fooled by the small CPU times, the work is being done on ms.lmfdb.xyz -- during the 5 minutes that this query was running, the mongod process was using 50% of the CPU, so something like 150s of CPU time to read through 57000 records, which is ridiculous.  I believe the CPU time is being spent decompressing the data.

How hard would it be to set up a test mongo db running wired tiger snappy?  I'd be curious to try the same query (I would only need enough space to copy over the forms collection, which is about 70GB).

We could probably ameliorate this particular problem by moving the hecke eigenvalue data out of the forms collection into a separate collection, but I suspect this issue is not specific to hmfs; there are other collections where similar issues may arise.

**Comment by AndrewVSutherland** [comment]
I just added the critical tag to this issue because I believe it is the cause of #1600 and the stackdriver warnings we have been getting.  This (and similar) queries put a heavy CPU load on ms.lmfdb.xyz that persists well after the 30s timeout of the lmfdb worker thread -- even after the user sees the server error in their web browser, mongodb is still grinding away trying to finish the query (so a simple DOS attack on the lmfdb is to simply issue the above query in a half dozen browser tabs at once).

I think we may need to revisit the decision to switch to zlib.

**Comment by AndrewVSutherland** [comment]
I updated https://github.com/LMFDB/lmfdb/issues/1599#issuecomment-228822188 to include the wall time for the query run on Atkin which did not get pasted in.  The key points is this: for this particular query wired tiger zlib appears to be more than 300 times slower than mmap.

**Comment by AndrewVSutherland** [comment]
@edgarcosta A possibly relevant paragraph from https://s3.amazonaws.com/info-mongodb-com/MongoDB-Performance-Best-Practices.pdf:

> With the WiredTiger storage engine, use of XFS is strongly
recommended to avoid performance issues that have been
observed when using EXT4 with WiredTiger.

I notice that we are running EXT4 on ms.lmfdb.xyz.

**Comment by JohnCremona** [comment]
This is relevant for the planned switchover to WiredTiger at Warwick at some point:  although we did once use XFS for the filesystem of the /data partition, we now use BTRFS.

**Comment by AndrewVSutherland** [comment]
@JohnCremona, @jvoight One thing that would ameliorate this problem (and might be a good idea in any case) is to move the hecke eigvenalues into a separate collection.  This would be very easy to do as far as the web server code goes (I think a few lines in web_HMF.py and the download functions in hilbert_modular_form.py are the only places where changes would be required).

But I'm reluctant to just jump in and do this because I notice that you also have code in there for importing and recalculating data that looks like it would also need to change.  Any thoughts?

**Comment by JohnCremona** [comment]
I'm hoping that there will be time for John and I to talk about this this week, but he'll also be heavily in demand for some real mathematics!   
My suggestion would be to use our new canonical prime ordering (which I explained to JV this afternoon) to do what you suggest but make the new collection have the eigenvalues in the new canonical order.  If we did this first (creating teh new eigenvalues collection, using newform labels to match data with exitsing newforms), then 2nd would be to adapt the code which imports to and reads from the database; and then finally we could rewrite the old newforms db without the old ap in there.  At that point the order list of primes would also not be needed in the fields collection.
This does need careful planning, of course!

**Comment by edgarcosta** [comment]
can you confirm that the ```Uptime check failed``` for 15 min was due to reindexing of HMF?

@AndrewVSutherland should we try the DB disk with XFS? I can do that in a couple of hours

**Comment by AndrewVSutherland** [comment]
@edgarcosta I believe it was a combination of that and testing that was putting a high cpu load on ms.lmfdb.xyz -- check the cpu chart on GCE and you will see the times line up.

Regarding XFS, yes I think that would be worth trying.

**Comment by edgarcosta** [comment]
FYI, In a couple of hours I will talk with Drew to discuss the problem, and hopefully we will come up with a decent solution.

ps: I'm in Australia at the moment.

**Comment by AndrewVSutherland** [comment]
An index on dimension,degree has been added on ms.lmfdb.xyz so that the specific query that identified this issue is now very quick.  But other queries may still be slow.  Edgar and I are running some experiments now with various combinations of storage engines and file systems.

**Comment by jvoight** [comment]
Thanks @AndrewVSutherland and @edgarcosta for dealing with this.  I noticed some slow searches before, and they were often fixed by adding an index.  

I'm focused on math this week and don't know about file systems, so I'll leave y'all to that.

This doesn't help in the short term, but in the medium term, the proposal mentioned by @JohnCremona  to make a new eigenvalues collection seems like a good one.  I'm not sure if it's worthy of 1.1, but probably we should separate it from this critical production problem.

**Comment by edgarcosta** [comment]
I have hitted a wall, and decided to ask for help at the mongodb mailing list.
https://groups.google.com/d/msg/mongodb-user/4IW3rvgeGh4/qwfgv9IrDAAJ

**Comment by AndrewVSutherland** [comment]
Just to summarize what Edgar and I found last night, this performance issue is indeed a result of the switch from mmap to Wired Tiger (with zlib compression).  This is unfortunate because there are a lot of advantages to Wired Tiger -- it manages disk space much more efficiently and for many operations it is faster, but queries that require a linear search through a non-trivial amount of data can be dramatically slower.  As Edgar said, we are looking to get some additional support on this issue, but my guess is that we may end up deciding to switch back to mmap (we could run Wired Tiger with the less cpu intensive default compression option (snappy), or even no compression at all, both of which would reduce the CPU load, but it would only make these queries slower and not solve the original problem).

Regardless of which storage engine we use, it would be good to split the eigenvalue data out of the hms.forms collection.  I'm actually amazed that mmap handles the current implementation as well as it does, but a lot of things would be faster and easier to manage if we had two collections.  I will create a separate issue for this.

**Comment by AndrewVSutherland** [comment]
I just submitted a PR that effectively addresses this issue in a very safe way -- rather than pulling eigenvalue data out of hmfs.forms, I simply created a new collection hmfs.forms.search that is an exact copy of hms.forms minus the eigenvalue data (it takes less than 1/400 of the space of hmfs.forms, so the overhead of having duplicate information is negligible). See #1615 for details.

This removes any immediate pressure to switch from wired tiger back to mmap, and makes it unnecessary to do anything in the near term about changing how eigenvalue data is stored in hmfs.
Once this PR is merged and pushed to prod I will close this issue -- we still need to make a decision about wired tiger versus mmap, but it would be good to wait for feedback from the mongodb support forums before making it.
